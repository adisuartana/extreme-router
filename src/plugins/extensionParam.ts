import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle parameters with a specific file extension.
 * Syntax: `:file.css`
 */
export const extensionParam: Plugin = () => {
  const id = 'extensionParam';
  const priority = 500;
  const syntax = ':file.css';

  const handler: PluginHandler = (segment) => {
    const match = /^:(?<paramName>[a-zA-Z0-9_-]+)\.(?<extension>.+)$/.exec(segment);
    if (!match || !match.groups || !match.groups.paramName || !match.groups.extension) {
      return null;
    }
    const paramName = match.groups.paramName;
    const extension = match.groups.extension;
    return {
      paramName,
      additionalMeta: {
        extension,
      },
      match({ urlSegment, params }) {
        if (!urlSegment.endsWith(extension)) return false;
        params[paramName] = urlSegment.slice(0, -(extension.length + 1));
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
