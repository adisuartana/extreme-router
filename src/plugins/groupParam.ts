import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle grouped parameters.
 * Syntax: `:paramName(a|b)`
 */
export const groupParam: Plugin = () => {
  const id = 'groupParam';
  const priority = 300;
  const syntax = ':paramName(a|b)';

  const handler: PluginHandler = (segment) => {
    const match = /^:(?<paramName>[a-zA-Z0-9_-]+)\((?<dynamicGroup>[^|)]+(\|[^|)]+)+)\)$/.exec(segment);
    if (!match || !match.groups || !match.groups.paramName || !match.groups.dynamicGroup) {
      return null;
    }
    const paramName = match.groups.paramName;
    const dynamicGroup = match.groups.dynamicGroup;
    const group = Object.fromEntries(dynamicGroup.split('|').map((g) => [g, g]));
    return {
      paramName,
      additionalMeta: { group },
      match({ urlSegment, params }) {
        if (!group || !group[urlSegment]) return false;
        params[paramName] = urlSegment;
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
