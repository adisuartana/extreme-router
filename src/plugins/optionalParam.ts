import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle optional parameters.
 * Syntax: `:paramName?`
 */
export const optionalParam: Plugin = () => {
  const id = 'optionalParam';
  const priority = 600;
  const syntax = ':paramName?';

  const handler: PluginHandler = (segment) => {
    const match = /^:(?<paramName>[a-zA-Z0-9_-]+)\?$/.exec(segment);
    if (!match || !match.groups || !match.groups.paramName) {
      return null;
    }
    const paramName = match.groups.paramName;
    return {
      paramName,
      override: true,
      match({ urlSegment, params }) {
        params[paramName] = urlSegment;
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
