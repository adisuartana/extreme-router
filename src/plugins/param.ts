import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle basic parameters.
 * Syntax: `:paramName`
 */
export const param: Plugin = () => {
  const id = 'param';
  const priority = 700;
  const syntax = ':paramName';

  const handler: PluginHandler = (segment) => {
    const match = /^:(?<paramName>[a-zA-Z0-9_-]+)$/.exec(segment);
    if (!match || !match.groups || !match.groups.paramName) {
      return null;
    }
    const paramName = match.groups.paramName;
    return {
      paramName,
      match({ urlSegment, params }) {
        params[paramName] = urlSegment;
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
