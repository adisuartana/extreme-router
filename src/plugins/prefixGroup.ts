import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle prefix groups.
 * Syntax: `prefix(a|b)`
 */
export const prefixGroup: Plugin = () => {
  const id = 'prefixGroup';
  const priority = 100;
  const syntax = 'prefix(a|b)';

  const handler: PluginHandler = (segment) => {
    const match = /^(?<staticName>[a-zA-Z0-9_.-]+)\((?<dynamicGroup>[^|)]+(\|[^|)]+)+)\)$/.exec(segment);
    if (!match || !match.groups || !match.groups.staticName || !match.groups.dynamicGroup) {
      return null;
    }
    const staticName = match.groups.staticName;
    const dynamicGroup = match.groups.dynamicGroup;
    const group = Object.fromEntries(dynamicGroup.split('|').map((g) => [staticName + g, staticName + g]));
    return {
      paramName: '',
      additionalMeta: { group },
      match({ urlSegment }) {
        return !!(group && group[urlSegment]);
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
