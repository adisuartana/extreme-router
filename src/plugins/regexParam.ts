import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle parameters with regex validation.
 * Syntax: `:paramName<\d+>`
 */
export const regexParam: Plugin = () => {
  const id = 'regexParam';
  const priority = 400;
  const syntax = ':paramName<\\d+>';

  const handler: PluginHandler = (segment) => {
    const match = /^:(?<paramName>[a-zA-Z0-9_-]+)<(?<regex>.+)>$/.exec(segment);
    if (!match || !match.groups || !match.groups.paramName || !match.groups.regex) {
      return null;
    }
    const paramName = match.groups.paramName;
    const regex = new RegExp(`^${match.groups.regex}$`);
    return {
      paramName,
      additionalMeta: {
        regex,
      },
      match({ urlSegment, params }) {
        const match = urlSegment.match(regex);
        if (!match) return false;
        params[paramName] = match[0];
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
