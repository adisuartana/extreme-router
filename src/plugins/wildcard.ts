import type { Plugin, PluginHandler } from '../types';

/**
 * Plugin to handle wildcard parameters.
 * Syntax: `*` or `:wildcardName*`
 */
export const wildcard: Plugin = () => {
  const id = 'wildcard';
  const priority = 800;
  const syntax = '*';

  const handler: PluginHandler = (segment) => {
    const match = /^(?:\*|:(?<wildcardName>[a-zA-Z0-9_-]+)\*)$/.exec(segment);
    if (!match) {
      return null;
    }
    const paramName = match.groups?.wildcardName ?? '*';
    return {
      paramName,
      wildcard: true,
      match({ urlSegments, index, params }) {
        let rest = urlSegments[index];
        const segmentsLength = urlSegments.length;
        for (let j = index + 1; j < segmentsLength; j++) {
          rest += '/' + urlSegments[j];
        }
        params[paramName] = rest || '';
        return true;
      },
    };
  };

  return {
    id,
    priority,
    syntax,
    handler,
  };
};
