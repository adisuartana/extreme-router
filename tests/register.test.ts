import { beforeEach, describe, it, expect } from 'vitest';
import TestExtreme from './test-util';
import { param } from '../src/plugins/param';

describe('Extreme Router - Register() API', () => {
  let testExtreme: TestExtreme;
  type TestStore = { storeId: 'test' };
  const testStore: TestStore = { storeId: 'test' };

  beforeEach(() => {
    testExtreme = new TestExtreme<TestStore>({ storeFactory: () => testStore });
    testExtreme.use(param);
  });

  it('should register static paths correctly', () => {
    testExtreme.register('/');
    testExtreme.register('/test');
    testExtreme.register('/test/test2');
    testExtreme.register('/test/test2/test3');

    expect(testExtreme.getStaticPathCache()).toMatchObject({
      '/': testStore,
      '/test': testStore,
      '/test/test2': testStore,
      '/test/test2/test3': testStore,
    });
  });
  it('should register dynamic paths correctly', () => {
    testExtreme.register('/:id');
    testExtreme.register('/:id/:test');
    testExtreme.register('/:id/:test/user');
    expect(testExtreme.getRoot()).toMatchObject({
      dynamicChildren: [
        {
          pluginMeta: {
            paramName: 'id',
          },
          store: testStore,
          dynamicChildren: [
            {
              pluginMeta: {
                paramName: 'test',
              },
              store: testStore,
              staticChildren: {
                user: {
                  store: testStore,
                },
              },
            },
          ],
        },
      ],
    });
  });
  it('should throw error if static path is already registered', () => {
    testExtreme.register('/test/test2/test3');
    expect(() => testExtreme.register('/test/test2/test3')).toThrowErrorMatchingSnapshot();
  });
  it('should throw error if param path is already registered', () => {
    testExtreme.register('/:id/:test');
    expect(() => testExtreme.register('/:id/:test')).toThrowErrorMatchingSnapshot();
  });
  it('should throw error if param is already registered but with different param name', () => {
    testExtreme.register('/:id/:test/user');
    expect(() => testExtreme.register('/:opa/:test/user')).toThrowErrorMatchingSnapshot();
  });
});
