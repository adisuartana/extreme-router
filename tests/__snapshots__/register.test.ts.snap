// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Extreme Router - Register() API > should throw error if param is already registered but with different param name 1`] = `[Error: Dynamic segment already exists: :opa (Plugin ID conflict: param)]`;

exports[`Extreme Router - Register() API > should throw error if param path is already registered 1`] = `[Error: Path already registered: /:id/:test]`;

exports[`Extreme Router - Register() API > should throw error if static path is already registered 1`] = `[Error: Path already registered: /test/test2/test3]`;
